ipcc-key-value/src/ioctl.rs

```diff
@@ -0,0 +1,133 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+//! IPCC `ioctl` interface.
+//!
+//! This module is tightly-coupled to the host OS image, and should only be used
+//! by services bundled with it (e.g., sled-agent and installinator).
+
+use crate::InstallinatorImageId;
+use crate::InstallinatorImageIdError;
+use crate::IpccKeyLookupError;
+use crate::PingError;
+use crate::Pong;
+use libc::c_int;
+use std::fs::File;
+use std::io;
+use std::os::unix::prelude::AsRawFd;
+
+pub struct Ipcc {
+    file: File,
+}
+
+impl Ipcc {
+    pub fn open() -> io::Result<Self> {
+        let file = File::options().read(true).write(true).open(IPCC_DEV)?;
+        Ok(Self { file })
+    }
+
+    pub fn ping(&self) -> Result<Pong, PingError> {
+        const EXPECTED_REPLY: &[u8] = b"pong";
+
+        let mut buf = [0; EXPECTED_REPLY.len()];
+        let n = self.key_lookup(IpccKey::Ping, &mut buf)?;
+        let buf = &buf[..n];
+
+        if buf == EXPECTED_REPLY {
+            Ok(Pong)
+        } else {
+            Err(PingError::UnexpectedReply(buf.to_vec()))
+        }
+    }
+
+    pub fn installinator_image_id(
+        &self,
+    ) -> Result<InstallinatorImageId, InstallinatorImageIdError> {
+        let mut buf = [0; InstallinatorImageId::CBOR_SERIALIZED_SIZE];
+        let n = self.key_lookup(IpccKey::InstallinatorImageId, &mut buf)?;
+        let id = InstallinatorImageId::deserialize(&buf[..n])
+            .map_err(InstallinatorImageIdError::DeserializationFailed)?;
+        Ok(id)
+    }
+
+    fn key_lookup(
+        &self,
+        key: IpccKey,
+        buf: &mut [u8],
+    ) -> Result<usize, IpccKeyLookupError> {
+        let mut kl = IpccKeyLookup {
+            key: key as u8,
+            buflen: u16::try_from(buf.len()).unwrap_or(u16::MAX),
+            result: 0,
+            datalen: 0,
+            buf: buf.as_mut_ptr(),
+        };
+
+        let result = unsafe {
+            libc::ioctl(
+                self.file.as_raw_fd(),
+                IPCC_KEYLOOKUP,
+                &mut kl as *mut IpccKeyLookup,
+            )
+        };
+
+        if result != 0 {
+            let errno = unsafe { *libc::___errno() };
+            return Err(IpccKeyLookupError::IoctlFailed { errno });
+        }
+
+        match kl.result {
+            IPCC_KEYLOOKUP_SUCCESS => Ok(usize::from(kl.datalen)),
+            IPCC_KEYLOOKUP_UNKNOWN_KEY => {
+                Err(IpccKeyLookupError::UnknownKey { key: format!("{key:?}") })
+            }
+            IPCC_KEYLOOKUP_NO_VALUE => Err(IpccKeyLookupError::NoValueForKey),
+            IPCC_KEYLOOKUP_BUFFER_TOO_SMALL => {
+                Err(IpccKeyLookupError::BufferTooSmallForValue)
+            }
+            _ => Err(IpccKeyLookupError::UnknownResultValue(kl.result)),
+        }
+    }
```

Comment:
The other variants should also include the key.

Reply:
Two thoughts:

* I was considering the key an internal implementation detail, and only exposed it in `UnknownKey` as an aid to debugging "your SP is too old" (which should be the only way we see that error).
* If you think it's valuable to expose that detail in more variants, do you think I should lift the known keys out to a `pub` enum instead of embedding them as debug strings?

Reply:
Ah yeah, I think printing it even for known keys might end up helping with debugging in the field. I think turning `IpccKey` into a `pub enum` would be ideal.

Reply:
Yeah, makes sense. Done in cd1728c